shared/api/v3/course.ts

```diff
@@ -0,0 +1,185 @@
+
+const timeRegex = /\d{2}:\d{2}/;
+const dateRegex = /\d{4}-\d{2}-\d{2}/;
+
+export const Schedule = z.object({
+    scheduleDays: z.string().regex(/[MTWRFSU]*/),
+    scheduleStartTime: z.string().regex(timeRegex),
+    scheduleEndTime: z.string().regex(timeRegex),
+    scheduleStartDate: z.string().regex(dateRegex),
+    scheduleEndDate: z.string().regex(dateRegex),
+    scheduleTermCount: z.number().int().positive(),
+    scheduleTerms: z.number().int().nonnegative().array(),
+    scheduleLocation: z.string(),
+});
+export type Schedule = z.infer<typeof Schedule>;
+
+export enum EnrollmentStatus {
+    open = "open",
+    closed = "closed",
+    reopened = "reopened",
+    unknown = "unknown",
+}
+export const EnrollmentStatusEnum = z.nativeEnum(EnrollmentStatus);
+export type EnrollmentStatusEnum = z.infer<typeof EnrollmentStatusEnum>;
+
+export const Course = z.object({
+    courseCode: z.string(),
+    courseName: z.string(),
+    courseSortKey: z.string().array(),
+    courseMutualExclusionKey: z.string().array(),
+    courseDescription: z.string(),
+    courseInstructors: z.string().array(),
+    courseTerm: z.string(),
+    courseSchedule: Schedule.array(),
+    courseCredits: z.number().nonnegative(),
+    courseSeatsTotal: z.number().int().nonnegative(),
+    courseSeatsFilled: z.number().int().nonnegative(),
+    courseWaitlistLength: z.null(),
+    courseEnrollmentStatus: EnrollmentStatusEnum,
+    permCount: z.number().int().nonnegative(),
+});
+export type Course = z.infer<typeof Course>;
+
+export function courseFromV4Section(s: APIv4.Section): Course {
+    const sectionCodeString = APIv4.stringifySectionCode(s.identifier);
+    const courseCodeString = APIv4.stringifyCourseCode(s.course.code);
+
+    const startDate = stringifyCourseDate(s.startDate);
+    const endDate = stringifyCourseDate(s.endDate);
+
+    const { termCount, terms } = getSectionTermFractions(s.identifier);
+
+    return {
+        courseCode: sectionCodeString,
+        courseName: s.course.title,
+        courseSortKey: [sectionCodeString],
+        courseMutualExclusionKey: [courseCodeString],
+        courseDescription: s.course.description,
+        courseInstructors: s.instructors.map((inst) => inst.name),
+        courseTerm: stringifySectionTermCode(s.identifier),
+        courseSchedule: s.schedules.flatMap((schedule) => {
+            const days = stringifyWeekdayEnumArray(schedule.days);
+            const startTime = stringifyTime(schedule.startTime);
+            const endTime = stringifyTime(schedule.endTime);
+            return schedule.locations.map((location) => ({
+                scheduleDays: days,
+                scheduleStartTime: startTime,
+                scheduleEndTime: endTime,
+                scheduleStartDate: startDate,
+                scheduleEndDate: endDate,
+                scheduleTermCount: termCount,
+                scheduleTerms: terms,
+                scheduleLocation: location,
+            }));
```

Comment:
You can inline all of the calls here to something like
```ts
return schedule.locations.map((location) => ({
                scheduleDays: schedule.days.join(""),
                scheduleStartTime: stringifyTime(schedule.startTime),
                scheduleEndTime: stringifyTime(schedule.endTime),
                scheduleStartDate: startDate,
                scheduleEndDate: endDate,
                scheduleTermCount: termCount,
                scheduleTerms: terms,
                scheduleLocation: location,
            }));
```

This increases both the readability of the code by not requiring the reader to jump back and forth between definitions and reducing runtime memory needed by not allocating extra once-off variables.

Reply:
The reason I didn't inline these values is because I didn't want it to instantiate new strings for every single element of `location`, since `days`, `startTime`, and `endTime` would be the same for each one. By factoring them out, they call all share a reference to the same string.